### PR TITLE
Delete Case Types and Properties in Data Dictionary V2

### PR DESCRIPTION
**Pull request by zandre-eng**
Exact same changes as [this PR](https://github.com/dimagi/commcare-hq/pull/33956), but with the commit history cleaned up a bit (no new code).

## Product Description

There is a new button to delete a case property. This button is contextual and will only appear for case properties that are not being used to store actual case data:
delete-case-prop-button

Clicking on the case property delete button will cause a confirmation modal to pop up:
Screenshot from 2024-02-07 14-55-23

After the user has clicked to confirm deletion, the case property will then be marked as deleted. The user will then need to confirm all changes by clicking "Save" at the top-right corner. This follows the same workflow as deprecating a case property.

Similarly, there is a new button to delete a case type. This button is also contextual and only appears if both the case type and all its case properties are not being used by cases:
delete-case-type-button

Clicking the case type delete button will cause a confirmation modal to pop up:
Screenshot from 2024-02-07 14-57-13

Once the user has confirmed to delete a case type, the page will be refreshed and the deleted case type will be gone from the left-hand side.

If there is an error with deleting a case type, a generic error message will be shown to the user:
delete-case-type-error

## Technical Summary

Link to ticket [here](https://dimagi-dev.atlassian.net/browse/SC-3076).

This adds in the ability for a user to delete case types/properties. It should be noted however, that this only applies to case types/properties that have no associated case data. Once a case starts using a specific case type/property, the type/property can then no longer be deleted in the Data Dictionary. The only way to achieve deleting it would be to first remove the relevant cases.

Note: This PR also contains the following PRs which have been merged into it:
- [Fix Delete Case Type/Prop Errors for Data Dictionary](https://github.com/dimagi/commcare-hq/pull/34073)
- [Change Case Property Delete Button](https://github.com/dimagi/commcare-hq/pull/34085)

The above PRs have been separately reviewed and approved.

## Feature Flag

None.

### Safety story

- Local testing.
- Has gone through QA.
- Automated tests.

### Automated test coverage

Automated tests have been created for the new view and util functions for deleting a case type. Furthermore, the existing unit tests for updating a case property have been updated to test for deleting case properties as well.

### QA Plan

QA has passed. Ticket available [here](https://dimagi-dev.atlassian.net/browse/QA-5997).

**Note:** Due to all the new changes/commits, QA will be doing a second pass on this feature. This will include doing some performance testing to ensure that the function responsible for collecting the list of case types/props that are safe to delete does not incur a big performance hit.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by kaapstorm — APPROVED** [review]
Reviewed just the last three commits, which are the only commits that have not yet been reviewed already.

**Review by Charl1996 — APPROVED** [review]
LGTM

**Comment by mkangia** [comment]
@zandre-eng thanks for a more compact PR

Just confirming that this PR has just two new changes after the closed PR? 
Since we changed the commits, I assume you have confirmed that changes to code are still the same as closed PR before those two new commits?

Reviewed everything after and including the commit "add new css class for auto margin" since I had reviewed everything before that in the closed PR. All looks good.

Just couple of non blocking comments for you to consider:
1. the deprecated icon now looks really small next to the delete button for case properties. It will almost get missed. I suggest we change that to a button as well. Having one icon and one button isn't looking great. Plus, that will be more in line with rest of HQ.
This can be taken up in a follow up PR if you wish to not block this PR and check with others about the change first.
2. when deleting case type, its an immediate action. We have changed that text to "Delete" (PR description still has the screenshot with "Confirm" text, so good to update that) which looks good. However, for case properties, it's not immediate. We should let the user know that we are just "marking" the case property for deletion and it will happen when they click save. You could consider simply changing the text in modal to say "Would you like to mark this case property to be deleted. Once confirmed the case property will be deleted on clicking save" and change the button text to "Confirm" instead.
Same not about taking in a follow up PR if needed.

Thanks for your effort on this feature change.

**Comment by zandre-eng** [comment]
> @zandre-eng thanks for a more compact PR
...

@mkangia That is correct, there are currently only two new commits which change the styling and text of the delete buttons. Otherwise, the code is all exactly the same. The commits might look a little different, but this is simply because I ended up squashing a few commits together from the previous PR.

Regarding your questions:
- I agree the button layout does look a little weird, however I was hesitant to just change the pre-existing deprecate button. I do also agree it would look better as a button, although I do think input from others would be valuable before changing it. Additionally, given the focus and number of changes already present in this PR, I'm hesitant to bloat it with too many additional changes. I can create a separate Jira ticket for this suggestion, which will give us a chance to discuss this change and keep this PR unblocked.
- Same as with the first question, I think this can be tackled in a separate PR. I can also put this into a new Jira ticket for discussion. Thanks for the note on the outdated screenshot, will update.

Please let me know if you're happy with the above answers. If so, I'll ahead and create the necessary Jira tickets.

**Comment by mkangia** [comment]
@zandre-eng 
yes absolutely, good to address those two points separately.

**Review by SmittieC — APPROVED** [review]

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 1]
```diff
@@ -220,7 +220,7 @@ hqDefine("data_dictionary/js/data_dictionary", [
                     }
 
                     _.each(group.properties(), function (element, index) {
-                        if (element.deleted && !element.id) {
+                        if (element.deleted() && !element.id) {
```
Comment (mkangia) [inline corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:223]: :+1: 

This got caught in QA?

Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:223]: This did. Creating case properties wasn't getting persisted on the back-end, which led to the discovery of this missing parenthesis. I think this bug slipped in when I was cleaning up the commit history and created a new PR.

**Pull request merged**